### PR TITLE
Fix handler interface and use real exponential backoff

### PR DESCRIPTION
**Pull request by brianstorti**
We were using the wrong arity for the `reject` method in the `ExponentialBackoffHandler`. When a message is being requeued, `Sneakers` calls it with 4 parameters instead of 3. I changed the tests for this handler to actually test all the possible responses (`reject`, `error`, `timeout` and `requeue`) to make sure it always works.

I also removed the `delays` options and added real exponential backoff, so the only thing the worker can specify is the maximum number of retries, with the `max_retries` options, that uses a default of 25.

cc/ @toreriklinnerud @abepetrillo

### CODE REVIEW

File: lib/sneakers_handlers/exponential_backoff_handler.rb [hunk 1]
```diff
@@ -110,13 +114,17 @@ def create_error_exchange!
...
     def create_retry_queue!(delay)
       @channel.queue("#{queue.name}.retry.#{delay}",
-        durable: durable_queues?,
-        arguments: {
-          :"x-dead-letter-exchange" => options[:exchange],
-          :"x-message-ttl" => delay * 1_000,
-          :"x-expires" => delay * 1_000 * 2
-        }
-      )
+                     durable: durable_queues?,
```
Comment (abepetrillo) [inline lib/sneakers_handlers/exponential_backoff_handler.rb]: I prefer the old spacing, bot a blocker though.

**Comment by abepetrillo** [comment]
:+1:

File: lib/sneakers_handlers/exponential_backoff_handler.rb (hunk 1 above)
Comment (brianstorti) [inline lib/sneakers_handlers/exponential_backoff_handler.rb]: 👍

**Comment by brianstorti** [comment]
@abepetrillo @toreriklinnerud 

Fixed a couple of other things as well:

* Removed the need to have an additional retry exchange, now we have just the primary exchange and the dlx;

* Simplified the routing key logic. We were manipulating the routing keys quite a lot to make it work the way we wanted, and it was a bit hard to follow what was going on and a couple of corner cases were not being covered. Now it works like this:

  - Primary queue is bound to a routing key with its own name;
  - Retry queues are defined using the primary queue name as their `x-dead-letter-routing-key`

And that covers everything. When a message fails, it's published to a retry queue. When the message expires in the retry queue it is dead-lettered using the `x-dead-letter-routing-key`, that is the primary queue name, so the message is executed again (and again).

* Fixed an issue with cached queue names.  When we create a new queue, `Bunny` stores its name in an internal cache.  The problem is that as we are creating ephemeral queues that can expire shortly after they are created, this cached queue may not exit anymore when we try to publish a second message to it. We are now removing the queue from the cache before trying to create it to guarantee that `Bunny` will always try to check if they exist, and when they don't, it will create them for us.

* Fixed an issue with shoveled messages. When we shovel a message using the management UI, what `RabbitMQ` does is to publish a message to the `default exchange`, using the queue name as the routing key. When that happened, as the retry queues were not bound to a routing key with the primary queue name, the message was long.

File: lib/sneakers_handlers/exponential_backoff_handler.rb [hunk 2]
```diff
@@ -70,53 +89,58 @@ def death_count(headers)
...
+      (count + 1) ** 2
+    end
+
+    # When we create a new queue, `Bunny` stores its name in an internal cache.
+    # The problem is that as we are creating ephemeral queues that can expire shortly
+    # after they are created, this cached queue may not exit anymore when we try to
+    # publish a second message to it.
+    # Removing queues from the cache guarantees that `Bunny` will always try
+    # to check if they exist, and when they don't, it will create them for us.
+    def clear_queues_cache
```
Comment (abepetrillo) [inline lib/sneakers_handlers/exponential_backoff_handler.rb]: Nice find!

**Comment by abepetrillo** [comment]
:+1:

**Comment by toreriklinnerud** [comment]
Great, this looks awesome. Just briefly looked over this, but will put it on my todo to have a proper look through it.

File: lib/sneakers_handlers/exponential_backoff_handler.rb [hunk 3]
```diff
@@ -70,53 +89,58 @@ def death_count(headers)
...
-        }
-      )
+      clear_queues_cache
+      channel.queue( "#{queue.name}.retry.#{delay}",
+         durable: options[:queue_options][:durable],
+         arguments: {
+           :"x-dead-letter-exchange" => options[:exchange],
+           :"x-dead-letter-routing-key" => queue.name,
+           :"x-message-ttl" => delay * 1_000,
+           :"x-expires" => delay * 1_000 * 2
```
Comment (toreriklinnerud) [inline lib/sneakers_handlers/exponential_backoff_handler.rb]: Wouldn't it be better to set x-expires to a fixed addition to the message-ttl, like 1 minute? I assume there is not much reason to keep it around after it has served it's purpose?

Also just to confirm, if new messages arrive say half way into the queue's expiry time, the queue is re-declared so the unused time starts counting again from 0?

Comment (brianstorti) [inline lib/sneakers_handlers/exponential_backoff_handler.rb]: As long as the `x-expires` value is greater than `x-message-ttl` it should be fine, so yeah, we could change it to be `x-mesage-ttl + 1 minute`.

And yes, the `x-expires` time means that the queue will expire after being unused for that time, so if a new message arrives, the time would be "reset".

Comment (toreriklinnerud) [inline lib/sneakers_handlers/exponential_backoff_handler.rb]: Perfect
